bundles/org.openhab.binding.boschshc/src/main/java/org/openhab/binding/boschshc/internal/discovery/BridgeDiscoveryParticipant.java

```diff
@@ -89,73 +101,159 @@ public String getServiceType() {
+            return publicInformation;
+        }
     }
 
     @Override
     public @Nullable ThingUID getThingUID(ServiceInfo serviceInfo) {
-        String ipAddress = discoverBridge(serviceInfo).shcIpAddress;
-        if (!ipAddress.isBlank()) {
-            return new ThingUID(BoschSHCBindingConstants.THING_TYPE_SHC, ipAddress.replace('.', '-'));
+        String ipAddress = getFirstIPAddress(serviceInfo);
+        if (ipAddress != null) {
+            @Nullable
+            PublicInformation publicInformation = getOrComputePublicInformation(serviceInfo, ipAddress);
+            if (publicInformation != null) {
+                String resolvedIpAddress = publicInformation.shcIpAddress;
+                return new ThingUID(BoschSHCBindingConstants.THING_TYPE_SHC, resolvedIpAddress.replace('.', '-'));
+            }
         }
         return null;
     }
 
-    protected PublicInformation discoverBridge(ServiceInfo serviceInfo) {
-        logger.trace("Discovering serviceInfo {}", serviceInfo);
-
-        if (serviceInfo.getHostAddresses() != null && serviceInfo.getHostAddresses().length > 0
-                && !serviceInfo.getHostAddresses()[0].isEmpty()) {
-            String address = serviceInfo.getHostAddresses()[0];
-            logger.trace("Discovering InetAddress {}", address);
-            // store all information for later access
-            bridgeInformation = getPublicInformationFromPossibleBridgeAddress(address);
+    protected @Nullable PublicInformation discoverBridge(String ipAddress) {
+        logger.debug("Attempting to contact Bosch Smart Home Controller at IP {}", ipAddress);
+        PublicInformation bridgeInformation = getPublicInformationFromPossibleBridgeAddress(ipAddress);
+        if (bridgeInformation != null && bridgeInformation.shcIpAddress != null
+                && !bridgeInformation.shcIpAddress.isBlank()) {
+            return bridgeInformation;
         }
 
-        return bridgeInformation;
+        return null;
     }
 
-    protected PublicInformation getPublicInformationFromPossibleBridgeAddress(String ipAddress) {
+    /**
+     * Attempts to send a HTTP request to the given IP address in order to determine
+     * if the device is a Bosch Smart Home Controller.
+     * 
+     * @param ipAddress the IP address of the potential Bosch Smart Home Controller
+     * @return a {@link PublicInformation} object if the bridge was successfully
+     *         contacted or <code>null</code> if the communication failed
+     */
+    protected @Nullable PublicInformation getPublicInformationFromPossibleBridgeAddress(String ipAddress) {
         String url = BoschHttpClient.getPublicInformationUrl(ipAddress);
-        logger.trace("Discovering ipAddress {}", url);
+        logger.trace("Requesting SHC information via URL {}", url);
         try {
             httpClient.start();
             ContentResponse contentResponse = httpClient.newRequest(url).method(HttpMethod.GET).send();
```

Comment:
On my network I get about 175 mDNS events in 5 minutes, over 100 of which are for the Bosch Smart Home Controller. I think it is too much to ask the controller over 100 times if it is still there in this short amount of time 😉

Thanks for the hint about the timeout. I thought about that as well and will add it in the next commit.

Reply:
If the requests would be cached for 5 minutes, the amount of request would be dramatically reduced without need for restart the binding.

Reply:
I could indeed implement time-based cache invalidation, but then I would like to use the Google Guava caches. Do you know if it's possible to add Guava as a dependency, and how?

On the other hand, I don't know if this is really necessary. The controller IP has to be provided in the bridge configuration, so it has to be pinned down anyway. If the bridge IP changes for some reason (which I would assume this happens rarely), we would see two bridges in the discovery and would have to choose the right one. And the "old" one would disappear after an openHAB or bundle restart.

When we use the add-on, we need the bridge before doing anything else, so we care about the discovery only for a very short amount of time. Once the bridge is added as thing we don't care about the bridge discovery anymore. So I'm not sure if it's worth ensuring that "old" bridge IPs are deleted immediately.

What do you guys think?

Reply:
Maybe a @openhab/add-ons-maintainers has an opinion on the Guava cache? I would rather perform 1 `not needed` http request per 5 mins over a binding restart for end-user experience.
But maybe that request can also be prevented if the event is matched with an allready defined bridge? (never done this before, but can;t you look it up in some registry?

And if that match is possible, it would also be possible to handle a changed ip-address for the same bridge event, i guess.

Reply:
I'm afraid it's not that straight-forward. We can not directly derive a thing UID from a mDNS event.

We get mDNS with different IP addresses for the same bridge. I don't know the background but the same bridge is both advertised under a `169.254.*.*` IP address and a local IP such as `192.168.*.*`. When we send a REST request to the bridge, we get a response that contains the actual local IP address (in both cases the `192.168.*.*`) address.

With my current implementation we would have two cache entries (with two different IP addresses, `169.254.*.*` and  `192.168.*.*`) with an object containing the same local IP address as value. If we now get more of the frequent mDNS events, we perform a cache lookup, which always results in the `192.168.*.*` address, thus resulting in the same thing UID.

I'm not 100% sure but I think openHAB will not add the bridge to the discovery inbox if there is already a thing with the proposed thing UID. If this assumption is correct the current implementation handles it, but we still need the cache for efficient "IP resolving".

Looking at the discovery results, I see that we have to solve another problem though: When a discovery result is created and passed to the openHAB framework, it gets an individual time to live (TTL) property. This is currently set to 1 hour. I guess it does not make sense to have an internal cache timeout of several minutes if the result is still visible in the openHAB UI for possibly another hour. Maybe our cache timeout should be the same as the TTL of the discovery result. Do you agree?

Reply:
I think this should be tested to be sure. Unfortunately i don't have such a device, so i can't inspect the mDNS events/behaviour.  I can also not fully asses the openHAB TTL, but i guess it is not a big problem if there is an orphan entry in openHAB as long as that does not prevent a new discovery from taking place. 

Maybe it is better to get the current fix in and from there see if it can be further improved.

Reply:
I verified that my assumption with the thing UIDs is correct. If a discovery result is created with an already existing thing UID, it is not displayed in the inbox/UI. If the user changes the UID when adding it, the discovery result is still shown in the inbox (but this is not a new behavior, in fact I guess every binding has this "problem" with any discovery result).

We could merge the code as it is. Discovering new controllers with new IPs will always be possible. We only have a minor problem if a controller disappears. It will still be shown in the inbox although it's "gone". After a restart the entry will be gone. But I think this does not happen often and is somewhat acceptable.

If you want I can enhance the cache with a time-based eviction strategy. But for that I would either like to use Google Guava Caches or `PassiveExpiringMap` from Apache Commons Collections 4. I prefer not to re-invent the wheel if well-developed, reusable implementations already exist. If the @openhab/add-ons-maintainers could give me a hint if and how these dependencies can be used I would be grateful 🙂

Thanks for your insights and your reviews 👍

Reply:
@jlaur do you know whether and how we can use Google Guava or Commons Collections 4?

Reply:
Did you already have a look if one of the cache implementation of openHAB Core works for you?

See: https://www.openhab.org/javadoc/latest/org/openhab/core/cache/package-summary

Reply:
Thank you @wborn 🙂 Looks like `ExpiringCacheMap` is what we are looking for :+1: I will try it out now.

Reply:
Works like a charm, thanks again @wborn :+1:
I set the expiry time to 10 minutes both for the cache entries and the openHAB discovery results in the inbox.
Code is now open for a final review.